src/ripple/basics/impl/Number.cpp

```diff
@@ -0,0 +1,754 @@
+        throw std::overflow_error("Number::root infinity");
+    }
+    if (f < Number{} && d % 2 == 0)
+        throw std::overflow_error("Number::root nan");
+    if (f == Number{})
+        return f;
+
+    // Scale f into the range (0, 1) such that f's exponent is a multiple of d
+    auto e = f.exponent() + 16;
+    auto const di = static_cast<int>(d);
+    auto ex = [e = e, di = di]()  // Euclidean remainder of e/d
+    {
+        int k = (e >= 0 ? e : e - (di - 1)) / di;
+        int k2 = e - k * di;
+        if (k2 == 0)
+            return 0;
+        return di - k2;
+    }();
+    e += ex;
+    f = Number{f.mantissa(), f.exponent() - e};  // f /= 10^e;
+    bool neg = false;
+    if (f < Number{})
+    {
+        neg = true;
+        f = -f;
+    }
+
+    // Quadratic least squares curve fit of f^(1/d) in the range [0, 1]
+    auto const D = ((6 * di + 11) * di + 6) * di + 1;
+    auto const a0 = 3 * di * ((2 * di - 3) * di + 1);
+    auto const a1 = 24 * di * (2 * di - 1);
+    auto const a2 = -30 * (di - 1) * di;
+    Number r = ((Number{a2} * f + Number{a1}) * f + Number{a0}) / Number{D};
+    if (neg)
+    {
+        f = -f;
+        r = -r;
+    }
+
+    //  Newton–Raphson iteration of f^(1/d) with initial guess r
+    //  halt when r stops changing, checking for bouncing on the last iteration
+    Number rm1{};
+    Number rm2{};
+    do
+    {
+        rm2 = rm1;
+        rm1 = r;
+        r = (Number(d - 1) * r + f / power(r, d - 1)) / Number(d);
+    } while (r != rm1 && r != rm2);
+
+    //  return r * 10^(e/d) to reverse scaling
+    return Number{r.mantissa(), r.exponent() + e / di};
+}
+
+Number
+root2(Number f)
+{
+    if (f == one)
+        return f;
+    if (f < Number{})
+        throw std::overflow_error("Number::root nan");
+    if (f == Number{})
+        return f;
+
+    // Scale f into the range (0, 1) such that f's exponent is a multiple of d
+    auto e = f.exponent() + 16;
+    if (e % 2 != 0)
+        ++e;
+    f = Number{f.mantissa(), f.exponent() - e};  // f /= 10^e;
+
+    // Quadratic least squares curve fit of f^(1/d) in the range [0, 1]
+    auto const D = 105;
+    auto const a0 = 18;
+    auto const a1 = 144;
+    auto const a2 = -60;
+    Number r = ((Number{a2} * f + Number{a1}) * f + Number{a0}) / Number{D};
+
+    //  Newton–Raphson iteration of f^(1/2) with initial guess r
+    //  halt when r stops changing, checking for bouncing on the last iteration
+    Number rm1{};
+    Number rm2{};
+    do
+    {
+        rm2 = rm1;
+        rm1 = r;
+        r = (r + f / r) / Number(2);
+    } while (r != rm1 && r != rm2);
+
+    //  return r * 10^(e/2) to reverse scaling
+    return Number{r.mantissa(), r.exponent() + e / 2};
+}
+
+// Returns f^(n/d)
+
+Number
+power(Number const& f, unsigned n, unsigned d)
+{
+    if (f == one)
+        return f;
+    auto g = std::gcd(n, d);
+    if (g == 0)
+        throw std::overflow_error("Number::power nan");
+    if (d == 0)
+    {
+        if (f == -one)
```

Comment:
The case `f == one` is already handled on line 731.  And `f == -one` must be handled separately because `power(1, 0, 0) == 1` but `power(-1, 0, 0)` is nan and should throw an exception.  Only if either `n` or `d` is not 0 (the exponent is not nan) should `power(-1, n, d) == 1`.

These corner cases were designed to be consistent with [Annex F of the C standard](https://cigix.me/c17#F.10.4.4), which is itself designed to be consistent with [ISO/IEC 60559](https://www.iso.org/standard/80985.html).